src/platform/linux/x11/mod.rs

```diff
@@ -523,6 +535,39 @@ impl EventsLoop {
     }
 }
 
+impl EventsLoopProxy {
+    pub fn wakeup(&self) -> Result<(), EventsLoopClosed> {
+        // Update the `EventsLoop`'s `pending_wakeup` flag.
+        let display = match (self.pending_wakeup.upgrade(), self.display.upgrade()) {
+            (Some(wakeup), Some(display)) => {
+                wakeup.store(true, atomic::Ordering::Relaxed);
+                display
+            },
+            _ => return Err(EventsLoopClosed),
+        };
+
+        // Push an event on the X event queue so that methods like run_forever will advance.
+        let mut xev = ffi::XClientMessageEvent {
+            type_: ffi::ClientMessage,
+            window: self.root,
+            format: 32,
+            message_type: 0,
+            serial: 0,
+            send_event: 0,
+            display: display.display,
+            data: unsafe { mem::zeroed() },
+        };
+
+        unsafe {
+            (display.xlib.XSendEvent)(display.display, self.root, 0, 0, mem::transmute(&mut xev));
```

Comment:
I recall trying to implement this at some point using self.root as the window and missing some wakeup events. I'll test again with your branch.